executor/join.go

```diff
@@ -257,6 +264,78 @@ func (e *HashJoinExec) prepare() error {
 	return nil
 }
 
+func (e *HashJoinExec) encodeRow(b []byte, row *Row) ([]byte, error) {
+	b = codec.EncodeVarint(b, int64(len(row.RowKeys)))
+	for _, rowKey := range row.RowKeys {
+		// Every rowKey is encoded with handle and idx in tableNames.
+		b = e.encodeRowKey(b, rowKey)
+	}
+	b, err := codec.EncodeValue(b, row.Data...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return b, nil
+}
+
+func (e *HashJoinExec) encodeRowKey(b []byte, rowKey *RowKeyEntry) []byte {
+	b = codec.EncodeVarint(b, rowKey.Handle)
+	for i, tn := range e.tableNames {
+		if tn == rowKey.TableName {
+			b = codec.EncodeVarint(b, int64(i))
+			return b
+		}
+	}
+	tblIdx := int64(len(e.tableNames))
+	b = codec.EncodeVarint(b, tblIdx)
+	e.tableMap[tblIdx] = rowKey.Tbl
+	e.tableNames = append(e.tableNames, rowKey.TableName)
+	return b
+}
+
+func (e *HashJoinExec) decodeRow(data []byte) (*Row, error) {
+	row := new(Row)
+	data, entryLen, err := codec.DecodeVarint(data)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for i := 0; i < int(entryLen); i++ {
+		var entry *RowKeyEntry
+		data, entry, err = e.decodeRowKey(data)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		row.RowKeys = append(row.RowKeys, entry)
+	}
+	values := make([]types.Datum, e.smallExec.Schema().Len())
+	err = codec.SetRawValues(data, values)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	err = decodeRawValues(values, e.smallExec.Schema())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	row.Data = values
```

Comment:
I'm confused with Row struct, it has a Data and a RowKeys fields, what's the difference?